### ISSUE TITLE
Shard rebalance not obeying cluster_concurrent_rebalance

### ISSUE THREAD

**Issue by gibrown** [elastic/elasticsearch#6141]
During full cluster restart relocating shards steadily increased and did not respond to setting cluster.routing.allocation.cluster_concurrent_rebalance. This behavior is pretty clearly cause by having disk.threshold_enabled set.

- 00:30: Relocating shards was around 86 causing us to get tight on disk space (a node ran out of space and we deleted a number of shards with rm before restarting it).
- 00:57: Set cluster_concurrent_rebalance to 10 (trying to limit relocating shards)
- 01:00 Set to 4 (after realizing it was already set to 6 in the config).
- 01:33 Set to 2 (in case something had changed and it was now per node not for the whole cluster)
- 02:00 Relocating shards hit its peak of ~140

![image](https://cloud.githubusercontent.com/assets/820871/2953053/8521b826-da4c-11e3-8b56-749eca93943a.png)

I'm guessing there is an interaction with disk threshold and/or shard allocation. We're seeing a lot of disk space taken up by old shards that haven't yet been cleaned up (note this is from 03:38):

![image](https://cloud.githubusercontent.com/assets/820871/2953176/4722b9f4-da50-11e3-8469-b661b9d9ff9f.png)

System config:
- 175 primary shards (plus 2 replicas per shard) = 525 shards total
- 36 data nodes (3 master nodes).
- Total disk space in cluster is 26TB. Total of all primaries is about 4.6TB (x3 = 13.8TB). Should be plenty of space.
- Some shards are pretty big. Max of about 45GB. Generally range from 15-40GB. 
- Just ran a full cluster restart to go from v0.90.12 to v1.1.1 (went pretty smoothly otherwise)
- Shard allocation awareness: DC = dfw, iad, sat; PARITY=0,1 (so 6 zones with 6 servers in each)

Relevant config settings:
```
cluster:
  routing:
    allocation:
      node_initial_primaries_recoveries: 8
      node_concurrent_recoveries: 15
      cluster_concurrent_rebalance: 6
      awareness.attributes: dc, parity
      disk.threshold_enabled: true

indices:
  recovery:
    max_bytes_per_sec: 100mb
    concurrent_streams: 5
```

```
{
  "cluster_name" : "es_glbl_cluster",
  "status" : "yellow",
  "timed_out" : false,
  "number_of_nodes" : 39,
  "number_of_data_nodes" : 36,
  "active_primary_shards" : 175,
  "active_shards" : 469,
  "relocating_shards" : 135,
  "initializing_shards" : 56,
  "unassigned_shards" : 0
}
```

**Comment by dakrone** [comment]
@gibrown do you know if this is still an issue for you guys? I know we have fixed some issues in the disk threshold decider with regard to taking shard sizes into account, and I am wondering if this might also have been related. Have you seen this again recently?

**Comment by gibrown** [comment]
We've had the disk threshold allocation disabled since this incident so we wouldn't run into it.

We're not currently on 1.4.x yet either, but are planning an upgrade in a month or two. I'll add trying the disk allocation to our list.

**Comment by bsandvik** [comment]
When upgrading from 1.4.2 to 1.4.4 I observed that the settings we have in elasticsearch.yml related to allocation are not being honored:

```
cluster.routing.allocation.cluster_concurrent_rebalance: 16
cluster.routing.allocation.node_concurrent_recoveries: 16
```

After a node restart when ES decided it was time to rebalance the cluster, I only saw two concurrent rebalancings going on. But when I manually set it to 16 using the API

```
curl -XPUT localhost:9200/_cluster/settings -d '{"transient":{"cluster.routing.allocation.cluster_concurrent_rebalance":16}}'

**Comment by clintongormley** [comment]
@bsandvik did you just change those settings in the config file on an already running cluster?  Or did you do a full cluster restart?

**Comment by bsandvik** [comment]
@clintongormley I observed this during a rolling restart of the cluster.

**Comment by clintongormley** [comment]
@bsandvik ok - so existing cluster settings where taken into account, rather than new settings from config files.  what you describe makes sense.

**Comment by sawyercade** [comment]
Bringing some attention back to this bug, we're experiencing the same issue as well. We set cluster_concurrent_rebalance to 0 using the cluster settings api, and relocations are still being allowed to happen. What's really interesting is that the cluster_concurrent_rebalance setting is being checked both in the allocator (I'm using the datarank/tempest allocator plugin) and in the ConcurrentRebalanceAllocationDecider, and neither of them seem to be receiving the updated transient value, despite the following:

```
$ curl -XGET localhost:9200/_cluster/settings
{"persistent":{},"transient":{"cluster":{"routing":{"allocation":{"enable":"all","cluster_concurrent_rebalance":"0"}}}}}

```

/_cat/shards shows 4 of our shards started relocating well after we updated the setting.
